src/Core/PluginApplicationInterface.php

```diff
@@ -0,0 +1,80 @@
+<?php
+/**
+ * CakePHP(tm) : Rapid Development Framework (https://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ * @link          https://cakephp.org CakePHP(tm) Project
+ * @since         3.6.0
+ * @license       https://opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\Core;
+
+use Cake\Event\EventDispatcherInterface;
+use Cake\Event\EventManagerInterface;
+
+/**
+ * Interface for Applications that leverage plugins
+ */
+interface PluginApplicationInterface extends EventDispatcherInterface
+{
+    /**
+     * Add a plugin to the loaded plugin set.
+     *
+     * @param string|\Cake\Core\PluginInterface $name The plugin name or plugin object.
+     * @param array $config The configuration data for the plugin if using a string for $name
+     * @return $this
+     */
+    public function addPlugin($name, array $config = []);
+
+    /**
+     * Run bootstrap logic for loaded plugins
+     *
+     * @return void
+     */
+    public function pluginBootstrap();
+
+    /**
+     * Run routes hooks for loaded plugins
+     *
+     * @param \Cake\Routing\RouteBuilder $routes The route builder to use.
+     * @return \Cake\Routing\RouteBuilder
+     */
+    public function pluginRoutes($routes);
+
+    /**
+     * Run middleware hooks for plugins
+     *
+     * @param \Cake\Http\MiddlewareQueue $middleware The MiddlewareQueue to use.
+     * @return \Cake\Http\MiddlewareQueue
+     */
+    public function pluginMiddleware($middleware);
+
+    /**
+     * Run console hooks for plugins
+     *
+     * @param \Cake\Console\CommandCollection $commands The CommandCollection to use.
+     * @return \Cake\Console\CommandCollection
+     */
+    public function pluginConsole($commands);
+
+    /**
+     * Run events hook for plugins
+     *
+     * @param \Cake\Event\EventManagerInterface $eventManager Event manager instance.
+     * @return \Cake\Event\EventManagerInterface
+     */
+    public function pluginEvents(EventManagerInterface $eventManager);
+
+    /**
+     * Application hook for attaching events to the Application event manager instance.
+     *
+     * @param \Cake\Event\EventManagerInterface $eventManager Event manager instance.
+     * @return \Cake\Event\EventManagerInterface
+     */
+    public function events(EventManagerInterface $eventManager);
```

Comment:
Since now `PluginApplicationInterface` extends `EventDispatcherInterface` would it make sense to remove argument and return form `events ` and `pluginEvents` method? App that supports event hooks now has it's own event manager and listeners should be attached to that instance anyway.

```php
// App\Application.php
public function events()
{
    $this->getEventManager()->on(new MyListener);
}
```

and

```php
public function pluginEvents()
{
    $events = $this->getEventManager();
    foreach ($this->plugins->with('events') as $plugin) {
        $events = $plugin->events($events); //this stays as it is, plugin doesn't dispatch events
    }
    $this->setEventManager($events);
}
```

This way event handling could be even more encapsutaled.

Reply:
I'm ok with that. I added the argument to make the routes, events, and middleware hooks consistent in that they all get a 'thing' and return that 'thing'.  Making them 'closed' would add inconsistency with other hooks but simplify framework code.

Reply:
Yeah, I know I’ve been advocating for consistency but now with the current API this makes little sense to have event hooks get and return something since it’s going to be always an app’s dependency.